src/app/child-dev-project/reporting/query.service.spec.ts

```diff
@@ -0,0 +1,384 @@
+      "twoDaysAgoEventWithoutRelation"
+    );
+    twoDaysAgoEventWithoutRelation.date = moment().subtract(2, "days").toDate();
+    twoDaysAgoEventWithoutRelation.addChild(femaleChristianChild.getId());
+    twoDaysAgoEventWithoutRelation.getAttendance(
+      femaleChristianChild.getId()
+    ).status = absentAttendanceStatus;
+    twoDaysAgoEventWithoutRelation.addChild(maleChristianChild.getId());
+    twoDaysAgoEventWithoutRelation.getAttendance(
+      maleChristianChild.getId()
+    ).status = presentAttendanceStatus;
+
+    mockEntityMapper.loadType.and.callFake((entityClass) => {
+      switch (entityClass as any) {
+        case Child:
+          return Promise.resolve([
+            femaleMuslimChild,
+            femaleChristianChild,
+            maleChristianChild,
+            maleChild,
+          ]);
+        case School:
+          return Promise.resolve([privateSchool, normalSchool]);
+        case RecurringActivity:
+          return Promise.resolve([
+            normalActivity,
+            privateActivity,
+            activityWithoutLink,
+          ]);
+        case EventNote:
+          return Promise.resolve([
+            threeDaysAgoPrivateEvent,
+            twoWeeksAgoPrivateEvent,
+            twoDaysAgoEventWithoutRelation,
+            todayEventWithoutSchool,
+            sixDaysAgoNormalEvent,
+          ]);
+        case ChildSchoolRelation:
+          return Promise.resolve([
+            maleChildAttendsPrivateSchool,
+            femaleChristianChildAttendsNormalSchool,
+            maleChristianChildAttendsNormalSchool,
+          ]);
+        default:
+          return Promise.resolve([]);
+      }
+    });
+    service.loadData();
+  });
+
+  it("should be created", () => {
+    expect(service).toBeTruthy();
+  });
+
+  it("should return all children with specified attributes", async () => {
+    const maleChristianQuery = `${Child.ENTITY_TYPE}:toArray[*gender=M & religion=christian]`;
+    const maleChristians = await service.queryData(maleChristianQuery);
+    expect(maleChristians).toEqual([maleChristianChild]);
+
+    const femaleQuery = `${Child.ENTITY_TYPE}:toArray[*gender=F]`;
+    const females = await service.queryData(femaleQuery);
+    expect(females).toEqual(
+      jasmine.arrayWithExactContents([femaleChristianChild, femaleMuslimChild])
+    );
+
+    const allChildren = await service.queryData(`${Child.ENTITY_TYPE}:toArray`);
+    expect(allChildren).toEqual(
+      jasmine.arrayWithExactContents([
+        maleChristianChild,
+        maleChild,
+        femaleChristianChild,
+        femaleMuslimChild,
+      ])
+    );
+  });
+
+  it("should return all children attending a school based on attributes", async () => {
+    const maleChildrenOnPrivateSchoolsQuery = `
+      ${School.ENTITY_TYPE}:toArray[*privateSchool=true]
+      :getRelated(${ChildSchoolRelation.ENTITY_TYPE}, schoolId)
+      :getActive.childId:addPrefix(${Child.ENTITY_TYPE}):unique
```

Comment:
this chain is really hard to understand (`:getActive.childId:addPrefix(${Child.ENTITY_TYPE}):unique`) can we change the phrasing to make it more intuitive?

Reply:
What would you find more intuitive? Maybe `getAttendingChildren` and `removeDuplicates` but for `addPrefix` I can not think of a more explanatory name